lms/djangoapps/teams/csv.py

```diff
@@ -19,3 +27,199 @@ def load_team_membership_csv(course, response):
+        dupe_set = set()
+        for teamset_id in self.teamset_ids:
+            if teamset_id in dupe_set:
+                self.validation_errors.append("Teamset with id " + teamset_id + " is duplicated.")
+                return False
+            dupe_set.add(teamset_id)
+            if teamset_id not in teamset_ids:
+                self.validation_errors.append("Teamset with id " + teamset_id + " does not exist.")
+                return False
+            self.user_ids_by_teamset_id[teamset_id] = {m.user_id for m in CourseTeamMembership.objects.filter
+                                                       (team__course_id=self.course.id, team__topic_id=teamset_id)}
+        return True
+
+    def validate_user_enrolled_in_course(self, user):
+        """
+        Invalid states:
+            user not enrolled in course
+        """
+        if not CourseEnrollment.is_enrolled(user, self.course.id):
+            self.validation_errors.append('User ' + user.username + ' is not enrolled in this course.')
+            return False
+
+        return True
+
+    def validate_user_assignment_to_team_and_teamset(self, row):
+        """
+        Validates a user entry relative to an existing team.
+        row is a dictionary where key is column name and value is the row value
+        [andrew],masters,team1,,team3
+        [joe],masters,,team2,team3
+        """
+        user = row['user']
+        for teamset_id in self.teamset_ids:
+            team_name = row[teamset_id]
+            if not team_name:
+                continue
+            try:
+                # checks for a team inside a specific team set. This way team names can be duplicated across
+                # teamsets
+                team = CourseTeam.objects.get(name=team_name, topic_id=teamset_id, course_id=self.course.id)
+            except CourseTeam.DoesNotExist:
+                # if a team doesn't exists, the validation doesn't apply to it.
+                all_teamset_user_ids = self.user_ids_by_teamset_id[teamset_id]
+                error_message = 'The user {0} is already a member of a team inside teamset {1} in this course.'.format(
+                    user.username, teamset_id
+                )
+                if user.id in all_teamset_user_ids and self.add_error_and_check_if_max_exceeded(error_message):
+                    return False
+                else:
+                    self.user_ids_by_teamset_id[teamset_id].add(user.id)
+                    continue
+            max_team_size = self.course_module.teams_configuration.default_max_team_size
+            if max_team_size is not None and team.users.count() >= max_team_size:
+                if self.add_error_and_check_if_max_exceeded('Team ' + team.team_id + ' is already full.'):
+                    return False
+            if CourseTeamMembership.user_in_team_for_course(user, self.course.id, team.topic_id):
```

Comment:
Same idea, here: this will get called (rows * columns) times.  You could instead make one query for all team memberships in the course and use that to make a dict that maps (user, teamset_id) -> list of team ids, and if this (user, teamset_id) maps to a non-empty list, record the validation error and return false.

Reply:
Also, does this method catch the case where there are duplicate rows for the same user in the CSV?

Reply:
updated